src/physics/wind.f90

```diff
@@ -279,19 +331,150 @@ subroutine update_winds(domain, options)
+            endif
+        enddo
+        
+        !Compute relative correction factors for U and V based on input speeds
+        U_cor = ABS(domain%u%data_3d(ims:ime,:,jms:jme))/ &
+                (ABS(domain%u%data_3d(ims:ime,:,jms:jme))+ABS(domain%v%data_3d(ims:ime,:,jms:jme)))
+                
+        do k = kms,kme
+            corr_factor = ((sum(domain%advection_dz(ims,1:k,jms)))/domain%smooth_height)
+            !corr_factor = (k*1.0)/wind_k
+            corr_factor = min(corr_factor,1.0)
+            do i = ims,ime
+                do j = jms,jme
+                    domain%w%data_3d(i,k,j) = domain%w%data_3d(i,k,j) - corr_factor * (domain%w%data_3d(i,wind_k,j))
+                    
+                    !if ( (domain%u%data_3d(i,k,j)+domain%v%data_3d(i,k,j)) == 0) U_cor(i,k,j) = 0.5
+                enddo
+            enddo
+            ! Compute this now, since it wont change in the loop
+            ADJ_coef(:,k,:) = -2/domain%dx 
+        enddo
+        
+        !V_cor = 1 - U_cor 
+        
+
+        U_cor = 0.5
+        V_cor = 0.5
+        
+        ! Now, fixing w-winds, iterate over U/V to reduce divergence with new w-winds
+        do it = 0,options%parameters%wind_iterations
+            !Compute divergence in new wind field
+            call calc_divergence(div,domain%u%data_3d,domain%v%data_3d,domain%w%data_3d,domain%jacobian_u,domain%jacobian_v,domain%jacobian_w, &
+                                domain%advection_dz,domain%dx,domain%jacobian,domain%smooth_height)
+
+            !Compute adjustment based on divergence
+            ADJ = div/ADJ_coef
+        
+            !Distribute divergence among the U and V fields                                                                    
+            domain%u%data_3d(ims+2:ime,:,jms+1:jme-1) = domain%u%data_3d(ims+2:ime,:,jms+1:jme-1) + &
+                                                        (ADJ(ims+1:ime-1,:,jms+1:jme-1) * U_cor(ims+2:ime,:,jms+1:jme-1))
+                                                        
+            domain%u%data_3d(ims+2:ime,:,jms+1:jme-1) = domain%u%data_3d(ims+2:ime,:,jms+1:jme-1) - &
+                                                        (ADJ(ims+2:ime,:,jms+1:jme-1) * U_cor(ims+2:ime,:,jms+1:jme-1))
+                                                        
+            domain%v%data_3d(ims+1:ime-1,:,jms+2:jme) = domain%v%data_3d(ims+1:ime-1,:,jms+2:jme) + &
+                                                        (ADJ(ims+1:ime-1,:,jms+1:jme-1) * V_cor(ims+1:ime-1,:,jms+2:jme))
+                                                        
+            domain%v%data_3d(ims+1:ime-1,:,jms+2:jme) = domain%v%data_3d(ims+1:ime-1,:,jms+2:jme) - &
+                                                        (ADJ(ims+1:ime-1,:,jms+2:jme) * V_cor(ims+1:ime-1,:,jms+2:jme))
+            call domain%u%exchange_u()
+            call domain%v%exchange_v()
+            
+        enddo
+
+        !If an update loop, swap meta_data and data_3d fields back
```

Comment:
minor point, but since updates will be much more frequent than initial runs, it might be better to put the memory copy in the initialization instead of with the updates.